src/EventStore.ClientAPI/EventStoreCatchUpSubscription.cs

```diff
@@ -536,34 +536,34 @@ public class EventStoreStreamCatchUpSubscription : EventStoreCatchUpSubscription
         /// <returns></returns>
         protected override Task ReadEventsTillAsync(IEventStoreConnection connection, bool resolveLinkTos, UserCredentials userCredentials,
             long? lastCommitPosition, long? lastEventNumber) =>
-            ReadEventsInternalAsync(connection, resolveLinkTos, userCredentials, lastCommitPosition, lastEventNumber);
-
-        private async Task ReadEventsInternalAsync(IEventStoreConnection connection, bool resolveLinkTos,
-                       UserCredentials userCredentials, long? lastCommitPosition, long? lastEventNumber)
-        {
-            var slice = await connection.ReadStreamEventsForwardAsync(StreamId, _nextReadEventNumber, ReadBatchSize, resolveLinkTos, userCredentials).ConfigureAwait(false);
-            await ReadEventsCallbackAsync(slice, connection, resolveLinkTos, userCredentials, lastCommitPosition, lastEventNumber).ConfigureAwait(false);
-        }
-
-        private async Task ReadEventsCallbackAsync(StreamEventsSlice slice, IEventStoreConnection connection, bool resolveLinkTos,
-                       UserCredentials userCredentials, long? lastCommitPosition, long? lastEventNumber)
-        {
-            if (!(await ProcessEventsAsync(lastEventNumber, slice).ConfigureAwait(false)) && !ShouldStop)
-            {
-                await ReadEventsInternalAsync(connection, resolveLinkTos, userCredentials, lastCommitPosition, lastEventNumber).ConfigureAwait(false);
-            }
-            else
-            {
-                if (Verbose)
-                {
-                    Log.Debug(
-                        "Catch-up Subscription {0} to {1}: finished reading events, nextReadEventNumber = {2}.",
-                        SubscriptionName,
-                        IsSubscribedToAll ? "<all>" : StreamId,
-                        _nextReadEventNumber);
-                }
-            }
-        }
+            ReadEventsInternalAsync(connection, resolveLinkTos, userCredentials, lastEventNumber);
+
+	    private async Task ReadEventsInternalAsync(IEventStoreConnection connection, bool resolveLinkTos,
+		    UserCredentials userCredentials, long? lastEventNumber)
+	    {
+		    bool shouldStopOrDone;
+			do
+		    {
+			    var slice = await connection
+				    .ReadStreamEventsForwardAsync(StreamId, _nextReadEventNumber, ReadBatchSize, resolveLinkTos, userCredentials)
+				    .ConfigureAwait(false);
+				shouldStopOrDone = await ReadEventsCallbackAsync(slice, lastEventNumber).ConfigureAwait(false);
+		    } while (!shouldStopOrDone);
+	    }
+
+	    private async Task<bool> ReadEventsCallbackAsync(StreamEventsSlice slice, long? lastEventNumber)
+	    {
+		    bool shouldStopOrDone = ShouldStop || await ProcessEventsAsync(lastEventNumber, slice).ConfigureAwait(false);
```

Comment:
Here is one small 'functional' change. If at this point ShouldStop is true we don't call ProcessEventsAsync. Before it would first call ProcessEventsAsync and then check ShouldStop.
I think this is better because if you indicated that you want to stop, there is no need to process anymore events.